pullrequests/counting_bits/step2.go

```diff
@@ -0,0 +1,23 @@
+//lint:file-ignore U1000 Ignore all unused code
+package countingbits
+
+/*
+レビュワーの方へ：
+	- このコードは既にGoの標準のフォーマッタで整形済みです。演算子の周りにスペースがあったりなかったりしますが、これはGoのフォーマッタによるもので、優先順位の高い演算子の周りにはスペースが入らず、低い演算子の周りには入るようになっています。https://qiita.com/tchssk/items/77030b4271cd192d0347
+*/
+
+/*
+DPのメモカを利用すればO(n)で解ける。
+2の冪乗のときは必ず１になるし、それ以外の場合は前に計算した結果+1すれば良い。
+*/
+func countBitsStep2(n int) []int {
+	bitCounts := make([]int, n+1)
+	powerOfTwo := 1
+	for i := 1; i <= n; i++ {
+		if i == powerOfTwo*2 {
+			powerOfTwo = i
+		}
+		bitCounts[i] = 1 + bitCounts[i-powerOfTwo]
```

Comment:
> 2の冪乗のときは必ず１になるし、それ以外の場合は前に計算した結果+1すれば良い。

個人的には上記コメントをそのまま実装してある方が理解しやすいと思いました。上記コメントがあったから理解できましたが、コードだけではこの意図を汲み取れなかった気がします。（とくに `i == powerOfTwo*2` のときに `bitCounts[i] = 1 + bitCounts[i-powerOfTwo] = 1 + bitCounts[0] = 1` となる部分の理解に時間がかかりました）

```suggestion
		if i == powerOfTwo*2 {
		        bitCounts[i] = 1
			powerOfTwo *= 2
		} else {
  		        bitCounts[i] = 1 + bitCounts[i-powerOfTwo]
  		}
```

また細かいですが if 文中の `powerOfTwo = i` よりも `powerOfTwo *= 2` の方が次のべき乗になるイメージがわかりやすいように感じました。